enhancements/ingress/tunable-router-buffer-sizes.md

```diff
@@ -0,0 +1,165 @@
+---
+title: tunable-router-buffer-sizes
+authors:
+  - "@sgreene570"
+reviewers:
+  - "@Miciah"
+  - "@danehans"
+  - "@frobware"
+  - "@knobunc"
+approvers:
+  - "@knobunc"
+creation-date: 2020-08-18
+last-updated: 2020-08-18
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+# Tunable Router Header Buffer Sizes
+
+## Release Signoff Checklist
+
+- [X] Enhancement is `implementable`
+- [X] Design details are appropriately documented from clear requirements
+- [X] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This enhancement extends the IngressController API to allow the user to configure
+the size of the in-memory header buffers for an IngressController. By default, these values are
+set at fixed values to limit memory use for typical IngressController usage.
+
+## Motivation
+
+OpenShift users have applications that require large HTTP headers. HTTP requests with
+header lengths that exceed the size of the IngressController's maximum header buffer size are
+dropped by the IngressController. In some cases, this limitation blocks OCP users from upgrading 3.x
+to 4.x.
+
+### Goals
+
+1. Enable a cluster administrator to easily modify the HAProxy `tune.bufsize` & `tune.maxrewrite` values for an IngressController
+running on an OCP 4.x cluster.
+
+### Non-Goals
+
+1. Determine platform-specific default values for `tune.maxrewrite` and `tune.bufsize`. This is cumbersome, and not critical if cluster administrators
+can adjust the values themselves.
+
+## Proposal
+
+The [HAProxy tune.bufsize](https://cbonte.github.io/haproxy-dconv/1.7/configuration.html#tune.bufsize) setting specifies
+the size of per-session HAProxy HTTP message buffers. The [OpenShift Router](https://github.com/openshift/rotuer)
+sets this value to `32768` bytes by default.
+
+The [HAProxy tune.maxrewrite](https://cbonte.github.io/haproxy-dconv/1.7/configuration.html#tune.maxrewrite) setting specifies
+how much memory HAProxy will reserve from the buffer (with length `tune.bufsize`) for HTTP header rewriting and appending.
+The [OpenShift Router](https://github.com/openshift/rotuer) sets this value to `8192` bytes by default.
+
+The IngressController API is extended by adding an optional `HeaderBuffer` field with
+type `*IngressControllerHeaderBuffer` to `IngressControllerSpec`:
```

Comment:
Are these settings specific to HTTP?  If so, let's call the new API field `httpHeaderBuffer`.  If it isn't HTTP-specific though, we need to document the non-HTTP implications—for example, does this setting have an impact on the TLS handshake?

Reply:
These settings are for HTTP only, AIUI